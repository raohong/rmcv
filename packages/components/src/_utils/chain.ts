const chain = <T extends (...args: any[]) => void>(
  ...fns: (T | undefined)[]
): ((...args: Parameters<T>) => void) => {
  return (...args) => {
    fns.forEach((item) => {
      if (item) {
        item(...args);
      }
    });
  };
};

export default chain;
