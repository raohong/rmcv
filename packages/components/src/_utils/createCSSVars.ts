import kebabCase from 'lodash/kebabCase';

const createCSSVars = (
  prefix: string,
  style: Record<string, string | undefined | null>,
) => {
  const result = Object.entries(style).reduce((map, [key, value]) => {
    if (!value) {
      return map;
    }
    // @ts-ignore
    // eslint-disable-next-line no-param-reassign
    map[`--${prefix}-${kebabCase(key)}`] = value;

    return map;
  }, {} as React.CSSProperties);

  return result;
};

export default createCSSVars;
