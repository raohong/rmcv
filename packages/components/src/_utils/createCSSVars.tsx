import React from 'react';

export const createCSSVars = (vars: Record<string, number | string>) =>
  Object.entries(vars).reduce((result, [name, value]) => {
    // @ts-ignore
    result[`--${name}`] = value;
    return result;
  }, {} as React.CSSProperties);
