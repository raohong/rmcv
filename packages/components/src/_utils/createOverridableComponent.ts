import React, { RefAttributes } from 'react';

interface OverridableComponent<
  C extends React.ComponentType<React.ComponentProps<C>>,
  DefaultElement extends keyof JSX.IntrinsicElements = 'div',
  ExcludeProps extends object = object,
> extends React.FC<
    Omit<
      Omit<React.ComponentProps<C>, 'ref'> &
        Omit<
          JSX.IntrinsicElements[DefaultElement],
          Exclude<'ref', keyof React.ComponentProps<C>>
        >,
      keyof ExcludeProps | 'component'
    >
  > {
  // JSX.IntrinsicElements
  <Tag extends keyof JSX.IntrinsicElements>(
    props: Omit<
      Omit<
        JSX.IntrinsicElements[Tag],
        Exclude<'ref', keyof React.ComponentProps<C>>
      > &
        Omit<React.ComponentProps<C>, 'ref'>,
      keyof ExcludeProps | 'component'
    > & {
      component: Tag;
    },
  ): JSX.Element;

  // ClassComponent
  <Custom extends React.ComponentClass<React.ComponentProps<Custom>>>(
    props: Omit<
      Omit<React.ComponentProps<C> & React.ComponentProps<Custom>, 'ref'>,
      keyof ExcludeProps | 'component'
    > & {
      component: Custom;
    } & RefAttributes<InstanceType<Custom>>,
  ): JSX.Element;

  <
    Custom extends React.ForwardRefExoticComponent<
      React.ComponentProps<Custom>
    >,
  >(
    props: Omit<
      Omit<React.ComponentProps<C>, 'ref'> & React.ComponentProps<Custom>,
      keyof ExcludeProps | 'component'
    > & {
      component: Custom;
    },
  ): JSX.Element;
}

function createOverridableComponent<
  C extends React.ComponentType<React.ComponentProps<C>>,
  DefaultElement extends keyof JSX.IntrinsicElements = 'div',
  ExcludeProps extends object = object,
>(component: C): OverridableComponent<C, DefaultElement, ExcludeProps> {
  return component as unknown as OverridableComponent<
    C,
    DefaultElement,
    ExcludeProps
  >;
}

export default createOverridableComponent;
