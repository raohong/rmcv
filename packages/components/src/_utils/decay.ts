import { useEffect, useRef, useCallback } from 'react';
import type { SpringValue } from '@react-spring/web';
import raf from 'raf';

export type DecayAnimationOptions = {
  power?: number;
  timeConst?: number;
  from?: number;
  velocity?: number;
  modifyTarget?: (dist: number, velocity: number) => number;
  restDelta?: number;
};

type DecayAnimationState = {
  finished: boolean;
  value: number;
  velocity: number;
};

const decay = ({
  power = 0.8,
  timeConst = 300,
  from = 0,
  velocity = 0,
  restDelta = 0.2,
  modifyTarget,
}: DecayAnimationOptions) => {
  let movement = velocity * power * timeConst;
  const ideal = movement + from;

  const target = modifyTarget ? modifyTarget(ideal, velocity) : ideal;
  const state: DecayAnimationState = {
    finished: false,
    value: from,
    velocity,
  };

  if (target !== ideal) {
    movement = target - from;
  }

  return {
    next: (t: number) => {
      // y = a * Math.exp(kt)

      const dx = Math.exp((1 / -timeConst) * t);
      const delta = -movement * dx;

      state.finished = Math.abs(delta) < restDelta;
      state.value = state.finished ? target : target + delta;
      state.velocity = velocity * dx;

      return state;
    },
  };
};

export const runDecayAnimation = ({
  onChange,
  ...options
}: DecayAnimationOptions & {
  onChange: (state: DecayAnimationState) => void;
}) => {
  let id: number | null = null;
  let active = true;
  const startTime = Date.now();

  const generator = decay(options);

  const cancel = () => {
    if (id !== null) {
      raf.cancel(id);
      id = null;
    }

    active = false;
  };

  const run = () => {
    id = raf(() => {
      if (active) {
        const now = Date.now();
        const state = generator.next(now - startTime);

        onChange(state);

        if (!state.finished) {
          run();
        }
      }
    });
  };

  run();

  return cancel;
};

export const useDecayAnimation = () => {
  const unmountedRef = useRef(false);
  const canceller = useRef<(() => void)[]>([]);

  const cancelAnimation = useCallback(() => {
    if (canceller.current) {
      canceller.current.forEach((item) => item());
    }
  }, []);

  const runDecay = useCallback(
    (
      spring: SpringValue<number>,
      dir: number,
      boundary: number,
      options: DecayAnimationOptions,
    ) => {
      const cb = runDecayAnimation({
        ...options,
        onChange({ velocity, value }) {
          if (unmountedRef.current) {
            return;
          }

          if (
            (dir === 1 && value >= boundary) ||
            (dir === -1 && value <= boundary)
          ) {
            cb();
            canceller.current = canceller.current.filter((item) => item !== cb);

            spring.start({
              to: boundary,
              config: {
                decay: false,
                velocity,
              },
            });
          } else {
            spring.set(value);
          }
        },
      });

      canceller.current.push(cb);
    },
    [],
  );

  useEffect(() => {
    return () => {
      cancelAnimation();
    };
  }, [cancelAnimation]);

  return { runDecay, cancelAnimation };
};
