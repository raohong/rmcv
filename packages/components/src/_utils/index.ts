export { default as getPrefixCls } from './getPrefixCls';

export { default as isBrowser } from './isBrowser';

export { default as renderPortal } from './renderPortal';

export { default as uuid } from './uuid';

export { default as createCSSVars } from './createCSSVars';

export { default as chain } from './chain';

export { default as toArray } from './toArray';
