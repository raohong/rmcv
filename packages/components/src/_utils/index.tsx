export * from './createOverridableComponent';
export * from './decay';
export * from './flatReactNode';
export * from './getPrefixCls';
export * from './htmlAttrs';
export * from './reactRender';
export * from './useXId';
