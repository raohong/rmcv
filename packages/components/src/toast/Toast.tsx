import { isString, isPlainObject, isBrowser, omit } from '@rmc-vant/utils';
import type {
  ToastType,
  ToastConfig,
  ToastConfigType,
  ToastInstance,
  ToastOptions,
  ToastInterface,
} from './interface';
import createToastInstance from './createToastInstance';

let isMultiple = false;
let lastKey: string | null = null;
let toastInteface: ToastInterface | null = null;
let container: HTMLElement | null = null;
const defaultOptions = new Map<ToastConfigType, ToastConfig>();

const types: ToastType[] = ['fail', 'loading', 'normal', 'success'];
const configTypes: ToastConfigType[] = [...types, 'common'];

const santizeType = (type: ToastType = 'normal'): ToastType =>
  types.includes(type) ? type : 'normal';

const getOptions = (params: ToastType | ToastOptions) => {
  const baseOptions: ToastOptions = {
    ...defaultOptions.get('common'),
  };

  if (isString(params)) {
    baseOptions.type = santizeType(params);
    Object.assign(baseOptions, defaultOptions.get(baseOptions.type!));
  } else {
    Object.assign(baseOptions, {
      ...defaultOptions.get(baseOptions.type!),
      ...params,
      type: santizeType(params.type),
    });
  }

  return baseOptions;
};

function allowMultiple() {
  isMultiple = true;
}

function setDefaultOptions(type: ToastConfigType, options: ToastConfig): void;
function setDefaultOptions(options: ToastConfig & { type?: ToastConfigType }): void;
function setDefaultOptions(
  type: ToastConfigType | (ToastConfig & { type?: ToastConfigType }),
  options?: any,
) {
  const internalType = isPlainObject(type) ? type.type ?? 'common' : type;
  const internalOptions = isPlainObject(type) ? type : options;

  if (configTypes.includes(internalType)) {
    defaultOptions.set(internalType, { ...omit(internalOptions, ['type']) });
  }
}

function resetDefaultOptions(type?: ToastConfigType) {
  if (!type) {
    defaultOptions.clear();
  } else if (configTypes.includes(type)) {
    defaultOptions.delete(type);
  }
}

interface ToastWithInternalType {
  (message: string): ToastInstance;
  (options: Omit<ToastOptions, 'type'>): ToastInstance;
}

function Toast(message: string): ToastInstance;
function Toast(options: ToastOptions): ToastInstance;
function Toast(params: string | ToastOptions) {
  const ref: { current: ToastInstance | null } = {
    current: null,
  };

  if (!isBrowser) {
    return ref.current;
  }

  if (!toastInteface) {
    container = document.createElement('div');
    document.body.appendChild(container);
    toastInteface = createToastInstance(container);
  }

  const type = santizeType(isString(params) ? 'normal' : params.type || 'normal');

  const options = {
    ...getOptions(type),
  };

  let key: string | null = null;

  if (isString(params)) {
    options.message = params;
  } else {
    Object.assign(options, params);
    if ('loadingType' in params && params.type === undefined) {
      options.type === 'loading';
    }
  }

  if (toastInteface.instance.current) {
    key = toastInteface.instance.current.create(isMultiple, options);
    lastKey = key;

    ref.current = {
      update(nextOptions: ToastOptions) {
        toastInteface!.instance.current?.update(key!, nextOptions);
      },
      close: () => {
        toastInteface!.instance.current?.close(key!);
      },
    };
  }

  return ref.current;
}

function clear(clearAll?: boolean) {
  if (toastInteface?.instance?.current) {
    toastInteface.instance.current.close(
      clearAll ? undefined : lastKey ?? undefined,
    );
  }
}

function reset() {
  resetDefaultOptions();
  isMultiple = false;
  lastKey = null;
  toastInteface = null;

  if (container) {
    container.parentNode?.removeChild(container);
  }
}

type InternalAPIType = 'fail' | 'loading' | 'success';

interface IToastAPI extends Record<InternalAPIType, ToastWithInternalType> {
  resetDefaultOptions: typeof resetDefaultOptions;
  allowMultiple: typeof allowMultiple;
  setDefaultOptions: typeof setDefaultOptions;
  (message: string): ToastInstance;
  (options: ToastOptions): ToastInstance;
  clear: (clearAll?: boolean) => void;
  __reset: () => void;
}

function setAPI(api: IToastAPI) {
  const internalTypes: InternalAPIType[] = ['fail', 'loading', 'success'];

  internalTypes.forEach((type) => {
    api[type] = (params: string | ToastOptions) => {
      if (isString(params)) {
        return Toast({ type, message: params });
      }

      return Toast({
        ...params,
        type,
      });
    };
  });
}

const ToastAPI = Toast as IToastAPI;

ToastAPI.resetDefaultOptions = resetDefaultOptions;
ToastAPI.setDefaultOptions = setDefaultOptions;
ToastAPI.allowMultiple = allowMultiple;
ToastAPI.clear = clear;

if (process.env.NODE_ENV === 'test') {
  ToastAPI.__reset = reset;
}

setAPI(ToastAPI);

export default ToastAPI;
