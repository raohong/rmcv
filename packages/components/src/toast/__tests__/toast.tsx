import React from 'react';
import { render, screen } from '@testing-library/react';
import Toast from '../ToastComponent';
import { getPrefixCls } from '../../_utils';

const testId = 'test';

test('render correctly', () => {
  const tree = render(<Toast visible />);

  expect(tree.asFragment()).toMatchSnapshot();
});

test('render with visible', () => {
  const com = render(<Toast data-testid={testId} />);

  expect(screen.queryByTestId(testId)).not.toBeInTheDocument();

  com.rerender(<Toast visible data-testid={testId} />);

  expect(screen.queryByTestId(testId)).toBeInTheDocument();
});

test('render with position', () => {
  render(<Toast position="top" data-testid={testId} visible />);

  expect(screen.getByTestId(testId)).toHaveClass(getPrefixCls('toast-position-top'));
});

test('render with message', () => {
  const com = render(<Toast message="content" visible />);

  expect(com.container).toHaveTextContent('content');
});

test('render with type', () => {
  render(<Toast type="success" message="content" visible />);

  expect(screen.getByLabelText('Success')).toBeInTheDocument();
});

test('render with icon', () => {
  render(<Toast icon="icon" data-testid={testId} visible />);

  expect(screen.getByTestId(testId)).toHaveTextContent('icon');
});

test('render with closeOnClick', () => {
  const onClose = jest.fn();
  render(<Toast closeOnClick onClose={onClose} data-testid={testId} visible />);

  screen.getByTestId(testId).click();

  expect(onClose).toBeCalled();
});
