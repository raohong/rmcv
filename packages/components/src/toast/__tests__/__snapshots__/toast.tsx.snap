// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<DocumentFragment>
  <div
    aria-hidden="false"
    class="rmcv-popup rmcv-popup-center rmcv-popup-round rmcv-toast rmcv-toast-normal"
    style="z-index: 1002;"
  >
    <div
      class="rmcv-toast-message"
    />
  </div>
</DocumentFragment>
`;
